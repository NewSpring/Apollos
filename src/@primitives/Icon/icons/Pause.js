import React from 'react';
import PropTypes from 'prop-types';
import { Svg } from '../../Svg';
import makeIcon from './makeIcon';

const Icon = makeIcon(({ size = 32, fill, ...otherProps } = {}) => (
  <Svg width={size} height={size} viewBox="0 0 24 24" {...otherProps}>
    <Svg.Path
      d="M7 3h1.67v18H7V3zm8.33 0H17v18h-1.67V3z"
      fill={fill}
    />
  </Svg>
));

Icon.propTypes = {
  size: PropTypes.number,
  fill: PropTypes.string,
};

export default Icon;
