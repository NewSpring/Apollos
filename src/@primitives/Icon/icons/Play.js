import React from 'react';
import PropTypes from 'prop-types';
import { Svg } from '../../Svg';
import { iconFromSvgFont } from './makeIcon';

const Icon = iconFromSvgFont(({ size = 32, fill, ...otherProps } = {}) => (
  <Svg width={size} height={size} viewBox="0 0 512 512" {...otherProps}>
    <Svg.Path
      d="M125 478l0-441c0-3 1-4 1-4-2 3-5 6-9 8-5 2-9 3-12 2 0 0 1 1 3 2l274 214c5 4 5 5 0 9l-272 202c-2 2-4 2-4 2 3-1 7 0 12 2 4 2 6 5 8 8 0 0-1-1-1-4z m-40 0c0 33 22 44 49 24l272-202c26-19 27-52 1-72l-274-214c-26-21-48-11-48 23z"
      fill={fill}
    />
  </Svg>
));

Icon.propTypes = {
  size: PropTypes.number,
  fill: PropTypes.string,
};

export default Icon;
