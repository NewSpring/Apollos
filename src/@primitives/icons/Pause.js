import React from 'react';
import { Svg } from '../Svg';

export default ({ size = 32, fill = 'red' } = {}) => (
  <Svg width={size} height={size} viewBox="0 0 512 512">
    <Svg.Path
      d="M137 512c12 0 22-10 22-23l0-455c0-12-10-23-22-23-13 0-23 11-23 23l0 455c0 13 10 23 23 23z m227 0c13 0 23-10 23-23l0-455c0-12-10-23-23-23-12 0-23 11-23 23l0 455c0 13 11 23 23 23z"
      fill={fill}
    />
  </Svg>
);
