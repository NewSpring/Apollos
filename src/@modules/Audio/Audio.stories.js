import React from 'react';

import { storiesOf } from '@storybook/react-native';
import {
  View,
} from 'react-native';
import Play from '@primitives/icons/Play';
import Pause from '@primitives/icons/Pause';

import ThemeProvider from '@primitives/ThemeProvider';
import Audio from './';

storiesOf('Audio', module)
  // .add('renders', () => (
  //   <ThemeProvider>
  //     <Audio source="https://www.w3schools.com/html/horse.mp3">
  //       <Audio.Play>
  //         <View>
  //           <Play />
  //         </View>
  //       </Audio.Play>
  //
  //       <Audio.Pause>
  //         <View>
  //           <Pause />
  //         </View>
  //       </Audio.Pause>
  //
  //       <Audio.Seeker />
  //     </Audio>
  //   </ThemeProvider>
  // ));
