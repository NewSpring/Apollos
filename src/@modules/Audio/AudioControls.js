import React, { PureComponent } from 'react';
import { TouchableOpacity, View } from 'react-native';
import PropTypes from 'prop-types';
import { compose } from 'recompose';

import Play from '../../@primitives/icons/Play';
import Pause from '../../@primitives/icons/Pause';
import onlyPropTypes from '../../@utils/onlyPropTypes';

export class AudioControls extends PureComponent {
  static propTypes = {
    onPlay: PropTypes.func.isRequired,
    onPause: PropTypes.func.isRequired,
    isPlaying: PropTypes.bool.isRequired,
  };

  render() {
    const {
      onPlay,
      onPause,
      isPlaying,
      ...otherProps
    } = this.props;
    return (
      <TouchableOpacity
        onPress={isPlaying ? onPause : onPlay}
        {...otherProps}
      >
        <View>
          {isPlaying ? <Pause /> : <Play />}
        </View>
      </TouchableOpacity>
    );
  }
}

const enhance = compose(
  onlyPropTypes,
);

export default enhance(AudioControls);
