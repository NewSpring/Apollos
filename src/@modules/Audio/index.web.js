import React, { Component } from 'react';
import {
  View,
  TouchableHighlight,
} from 'react-native';
import PropTypes from 'prop-types';
import Play from '../../@primitives/icons/Play';
import Pause from '../../@primitives/icons/Pause';
import Seeker from '../../@primitives/Seeker';

export default class Audio extends Component {
  static propTypes = {
    source: PropTypes.string.isRequired,
    onReady: PropTypes.func,
    onError: PropTypes.func,
    onPlaybackReachedEnd: PropTypes.func,
    onPlay: PropTypes.func,
    onPause: PropTypes.func,
    onStop: PropTypes.func,
    onSeek: PropTypes.func,
    onSeeking: PropTypes.func,
  };

  static defaultProps = {
    onReady() {},
    onError() {
      // eslint-disable-next-line no-console
      console.error('Failed to load audio file');
    },
    onPlaybackReachedEnd() {},
    onPlay() {},
    onPause() {},
    onStop() {},
    onSeek() {},
    onSeeking() {},
  };

  state = {
    progress: 0,
  };

  componentDidMount() {
    this.audioFile.oncanplaythrough = () => {
      if (!this.isReady) this.props.onReady();
      this.isReady = true;
    };

    this.audioFile.onerror = () => {
      this.props.onError();
    };

    this.audioFile.onended = () => {
      this.props.onPlaybackReachedEnd();
    };

    this.audioFile.onplay = () => {
      this.props.onPlay();
    };

    this.audioFile.onpause = () => {
      this.props.onPause();
    };

    this.createStatusListener();
  }

  componentWillUnmount() {
    this.removeStatusListener();
  }

  get duration() {
    return this.audioFile && this.audioFile.duration;
  }

  positionListener = undefined;
  isReady = false;

  play = () => {
    if (this.isReady) this.audioFile.play();
  }

  pause = () => {
    this.audioFile.pause();
  }

  stop = () => {
    this.audioFile.pause();
    this.audioFile.currentTime = 0;
    this.props.onStop();
  }

  seek = (percentageOfSong) => {
    const positionInSeconds = this.duration * percentageOfSong;
    this.audioFile.currentTime = positionInSeconds;

    const positionInMillis = positionInSeconds * 1000;
    this.props.onSeek(positionInMillis);
  }

  handleSeeking = (percentageOfSong) => {
    const positionInSeconds = this.duration * percentageOfSong;
    const positionInMillis = positionInSeconds * 1000;
    this.props.onSeeking(positionInMillis);
  }

  createStatusListener = () => {
    this.positionListener = setInterval(() => {
      this.setState({
        progress: this.audioFile.currentTime / this.duration,
      });
    }, 200);
  }

  removeStatusListener = () => {
    if (this.positionListener) clearInterval(this.positionListener);
  }

  render() {
    const {
      source,
    } = this.props;
    return (
      <View>
        <audio
          ref={(r) => { this.audioFile = r; }}
        >
          <source src={source} type="audio/mpeg" />
        </audio>

        <TouchableHighlight onPress={this.play}>
          <View>
            <Play />
          </View>
        </TouchableHighlight>

        <TouchableHighlight onPress={this.pause}>
          <View>
            <Pause />
          </View>
        </TouchableHighlight>

        <Seeker
          progress={this.state.progress}
          onSeek={this.seek}
          onSeeking={this.handleSeeking}
        />
      </View>
    );
  }
}
