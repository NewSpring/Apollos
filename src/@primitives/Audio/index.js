import React, { Component } from 'react';
import {
  View,
  TouchableHighlight,
  PanResponder,
} from 'react-native';
import Play from '../icons/Play';
import Pause from '../icons/Pause';

export default class Audio extends Component {
  play = () => {
    this.audioFile.play();
  }

  pause = () => {
    this.audioFile.pause();
  }

  stop = () => {
    this.audioFile.currentTime = 0;
  }

  get duration() {
    return this.audioFile && this.audioFile.duration;
  }

  state = {
    progress: 0,
  };

  componentDidMount() {
    this.audioFile.canplaythrough = () => {
      console.log('is ready');
    };

    this.audioFile.onerror = () => {
      console.log('error');
    };

    this.audioFile.onplay = () => {
      console.log('play');
    };

    this.audioFile.onpause = () => {
      console.log('pause');
    };

    this.audioFile.onended = () => {
      console.log('ended');
    };

    this.positionListener = setInterval(() => {
      this.setState({
        progress: this.audioFile.currentTime / this.duration,
      });
    }, 200);
  }

  componentWillUnmount() {
    clearInterval(this.positionListener);
  }

  render() {
    return (
      <View>
        <audio
          ref={(r) => { this.audioFile = r; }}
        >
          <source src="https://www.w3schools.com/html/horse.mp3" type="audio/mpeg" />
        </audio>

        <TouchableHighlight onPress={this.play}>
          <View>
            <Play />
          </View>
        </TouchableHighlight>

        <TouchableHighlight onPress={this.pause}>
          <View>
            <Pause />
          </View>
        </TouchableHighlight>

        <View>
          <View style={{ backgroundColor: 'gray', height: 20 }} />
          <View style={{ backgroundColor: 'red', height: 20, position: 'absolute', left: 0, width: `${this.state.progress * 100}%` }} />
          <View style={{ backgroundColor: 'blue', borderRadius: 9999, position: 'absolute', left: `calc(${this.state.progress * 100}% - 5px)`, height: 10, width: 10 }} />
        </View>
      </View>
    );
  }
}
