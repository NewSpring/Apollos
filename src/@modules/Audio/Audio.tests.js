import React from 'react';
import renderer from 'react-test-renderer';
import {
  View,
} from 'react-native';
import ThemeProvider from '../../@primitives/ThemeProvider';
import Audio from './';
import Play from '../../@primitives/icons/Play';
import Pause from '../../@primitives/icons/Pause';

it('renders correctly', () => {
  const tree = renderer.create(
    <ThemeProvider>
      <Audio source="https://www.w3schools.com/html/horse.mp3">
        <Audio.Play>
          <View>
            <Play />
          </View>
        </Audio.Play>

        <Audio.Pause>
          <View>
            <Pause />
          </View>
        </Audio.Pause>

        <Audio.Seeker />
      </Audio>
    </ThemeProvider>,
  );
  // expect(true).toBe(true);
  expect(tree).toMatchSnapshot();
});
