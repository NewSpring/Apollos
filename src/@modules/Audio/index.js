import React, { Component } from 'react';
import { Audio as ExpoAudio } from 'expo';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import AudioControls from './AudioControls';
import AudioSeeker from './AudioSeeker';

export default class Audio extends Component {
  static propTypes = {
    source: PropTypes.string.isRequired,
    onReady: PropTypes.func,
    onError: PropTypes.func,
    onPlaybackReachedEnd: PropTypes.func,
    onPlay: PropTypes.func,
    onPause: PropTypes.func,
    onStop: PropTypes.func,
    onSeek: PropTypes.func,
    onSeeking: PropTypes.func,
    isPlaying: PropTypes.bool,
  };

  static defaultProps = {
    onReady() {},
    onError(err) {
      // eslint-disable-next-line no-console
      console.error(err.message);
    },
    onPlaybackReachedEnd() {},
    onPlay() {},
    onPause() {},
    onStop() {},
    onSeek() {},
    onSeeking() {},
    isPlaying: false,
  };

  static childContextTypes = {
    stop: PropTypes.func,
  };

  state = {
    isPlaying: this.props.isPlaying,
    progress: 0,
  };

  getChildContext = () => ({
    stop: this.stop,
  });

  componentWillMount() {
    this.Sound = new ExpoAudio.Sound();
    this.loadSource();
  }

  componentWillUnmount() {
    this.Sound.unloadAsync();
    this.removeStatusListener();
  }

  duration = 0;
  positionListener = undefined;
  previousSoundStatus = undefined;
  isReady = false;
  playbackEnded = false;

  play = async () => {
    try {
      if (this.playbackEnded) await this.Sound.stopAsync();
      if (this.isReady) {
        this.Sound.playAsync();
        this.props.onPlay();
        this.setState({
          isPlaying: true,
        });
      }
    } catch (err) {
      this.props.onError(err);
    }
  }

  pause = async () => {
    try {
      this.Sound.pauseAsync();
      this.props.onPause();
      this.setState({
        isPlaying: false,
      });
    } catch (err) {
      this.props.onError(err);
    }
  }

  stop = async () => {
    try {
      this.Sound.stopAsync();
      this.props.onStop();
    } catch (err) {
      this.props.onError(err);
    }
  }

  seek = (percentageOfSong) => {
    const positionInMillis = this.duration * percentageOfSong;
    this.Sound.setPositionAsync(positionInMillis);
    this.props.onSeek(positionInMillis);
  }

  handleSeeking = (percentageOfSong) => {
    const positionInMillis = this.duration * percentageOfSong;
    this.props.onSeeking(positionInMillis);
  }

  loadSource = async () => {
    const {
      source,
      onReady,
      onError,
    } = this.props;

    try {
      const soundStatus = await this.Sound.loadAsync({ uri: source });
      this.previousSoundStatus = soundStatus;
      this.duration = soundStatus.durationMillis;
      this.createStatusListener();
      this.isReady = true;
      onReady();
    } catch (err) {
      onError(err);
    }
  }

  createStatusListener = () => {
    this.positionListener = setInterval(async () => {
      try {
        const soundStatus = await this.Sound.getStatusAsync();
        this.setState({
          progress: soundStatus.positionMillis / this.duration,
        });

        const currentIsFinished = soundStatus.positionMillis === this.duration;
        const previousIsFinished = this.previousSoundStatus.positionMillis === this.duration;
        if (currentIsFinished && !previousIsFinished) {
          this.pause();
          this.props.onPlaybackReachedEnd();
        }
        this.previousSoundStatus = soundStatus;

        if (currentIsFinished) {
          this.playbackEnded = true;
        } else {
          this.playbackEnded = false;
        }
      } catch (err) {
        this.props.onError(err);
      }
    }, 200);
  }

  removeStatusListener = () => {
    if (this.positionListener) clearInterval(this.positionListener);
  }

  render() {
    return (
      <View>
        <AudioControls
          onPlay={this.play}
          onPause={this.pause}
          isPlaying={this.state.isPlaying}
        />
        <AudioSeeker
          progress={this.state.progress}
          seekingHandler={this.handleSeeking}
        />
      </View>
    );
  }
}
