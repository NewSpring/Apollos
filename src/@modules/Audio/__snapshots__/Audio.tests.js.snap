// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View>
      <View
        style={
          Array [
            Object {
              "transform": Array [
                Object {
                  "scaleY": -1,
                },
              ],
            },
          ]
        }
      >
        <Svg
          height={32}
          viewBox="0 0 512 512"
          width={32}
        >
          <Path
            d="M125 478l0-441c0-3 1-4 1-4-2 3-5 6-9 8-5 2-9 3-12 2 0 0 1 1 3 2l274 214c5 4 5 5 0 9l-272 202c-2 2-4 2-4 2 3-1 7 0 12 2 4 2 6 5 8 8 0 0-1-1-1-4z m-40 0c0 33 22 44 49 24l272-202c26-19 27-52 1-72l-274-214c-26-21-48-11-48 23z"
            fill="#6bac43"
          />
        </Svg>
      </View>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View>
      <View
        style={
          Array [
            Object {
              "transform": Array [
                Object {
                  "scaleY": -1,
                },
              ],
            },
          ]
        }
      >
        <Svg
          height={32}
          viewBox="0 0 512 512"
          width={32}
        >
          <Path
            d="M137 512c12 0 22-10 22-23l0-455c0-12-10-23-22-23-13 0-23 11-23 23l0 455c0 13 10 23 23 23z m227 0c13 0 23-10 23-23l0-455c0-12-10-23-23-23-12 0-23 11-23 23l0 455c0 13 11 23 23 23z"
            fill="#6bac43"
          />
        </Svg>
      </View>
    </View>
  </View>
  <View
    onLayout={[Function]}
  >
    <View
      style={
        Object {
          "backgroundColor": "#303030",
          "height": 20,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#6bac43",
          "height": 20,
          "left": 0,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
          "width": 0,
        }
      }
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "backgroundColor": "#1c683e",
          "borderRadius": 9999,
          "height": 30,
          "left": -15,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": -5,
            },
          ],
          "width": 30,
        }
      }
    />
  </View>
</View>
`;
